### ISSUE TITLE
negative transmat problem

### ISSUE THREAD

**Issue by huiyuHarvey** [hmmlearn/hmmlearn#108]
Hi, I am using `hmmlearn` to train a left-right hmm model with dozens of wav files.
```
self.pi_prior = np.array([1, 0.0, 0.0, 0.0, 0.0])
self.A_prior = np.array([[0.6, 0.4, 0.0, 0.0, 0.0],
                         [0.0, 0.5, 0.5, 0.0, 0.0],
                         [0.0, 0.0, 0.4, 0.6, 0.0],
                         [0.0, 0.0, 0.0, 0.5, 0.5],
                         [0.0, 0.0, 0.0, 0.0, 1.0]])

h = GMMHMM(n_components=5, n_mix=7, covariance_type = 'diag',
            startprob_prior=self.pi_prior, transmat_prior=self.A_prior, 
            n_iter=10, verbose=False, init_params='wmc', params='tmcw')

h.startprob_ = self.pi_prior
h.transmat = self.A_prior
```

However, the result after training is
```
 [[ 11.62414007 -10.62414007   0.          0.           0.        ]
 [ -0.           0.1013012    0.8986988   -0.          -0.        ]
 [  0.           0.           0.36092202   0.63907798   0.        ]
 [  0.           0.           0.          -0.38736272   1.38736272]
 [  0.           0.           0.           0.           1.        ]]
```
1) I don't get it. why there're negative elements in the `transmat` ?
2) Also, if I set the `init_params` to empty `' '`, it will raise the `ValueError` in line 607 `hmm.py`.
     I suppose this is because the `sklearn.mixture.GMM` module, right?

**Comment by superbobry** [comment]
>  I don't get it. why there're negative elements in the transmat ?

Hi, the negative transmat issue appeared in a Stack Overflow question recently, see my [answer](http://stackoverflow.com/a/35689457/262432). To avoid this you need to ensure `startprob_prior` and `transmat_prior` are >=1 for all states. 

> I suppose this is because the sklearn.mixture.GMM module, right?

Can you add the code and full traceback? I can't reproduce this on a trivial example:

```python
>>> hmm.GMMHMM(3, init_params="").fit([[1], [2], [3]])
GMMHMM(algorithm='viterbi', covariance_type='diag', covars_prior=0.01,
    init_params='', n_components=3, n_iter=10, n_mix=1, params='stmcw',
    random_state=None, startprob_prior=1.0, tol=0.01, transmat_prior=1.0,
    verbose=False)
```

**Comment by huiyuHarvey** [comment]
Thanks for the answer! 
So in my case, each element +1? if yes, how can I train a left-right GMM-HMM model? Besides, the `_check()` in `base.py` needs to be updated I guess. 
If I need to train on dozens of wav. files, which means to use the output `transmat_` as the input for next training, still there's chance some of the elements of `transmat_` may <= 1, how can I solve this?

The 2nd question in my first comment happened during the training of the 2nd file. The `covars_` of `gmm` start to drop below 0 when the error occurs.
The traceback is as follow:
```
    h.fit(cc)

  File "D:\Anaconda\lib\site-packages\hmmlearn\base.py", line 439, in fit
    bwdlattice)

  File "D:\Anaconda\lib\site-packages\hmmlearn\hmm.py", line 605, in _accumulate_sufficient_statistics
    raise ValueError

ValueError
```
The code is as below. The `_init_hmm_para()` is in the first comment.
```
    def train(self):
        h = self._init_hmm_para()
        flst = self._add_direct()
        for f in flst:
            print f
            audio = _input_audio(f)
            audio1 = _get_first_channel(audio)
            frames = _audio2frame(audio1)
            cc = _compute_cepstrum_coefficient(frames)

            h.fit(cc)
            pi = h.startprob_
            A = h.transmat_
        
        h = joblib.dump(h, "hmmgmm_model.pkl")
        return h, pi, A
```

**Comment by superbobry** [comment]
> If I need to train on dozens of wav. files, which means to use the output transmat_ as the input for next training, still there's chance some of the elements of transmat_ may <= 1, how can I solve this?

You don't need `transmat_prior` to preserve the transmat between sequences. Here's what you can do insted:

* Use transmat from the previous sequence as a starting point for the next sequence. To do this drop `"t"` from `init_params` and set the `transmat_` attribute to the previously estimated one. 
* Fit the model to multiple sequences using the `length` argument. See "Working with multiple sequences" section in the [tutorial](http://hmmlearn.readthedocs.org/en/latest/tutorial.html#training-hmm-parameters-and-inferring-the-hidden-states).

Please give any of these a try and let me know if it worked.

> The 2nd question in my first comment happened during the training of the 2nd file. The covars_ of gmm start to drop below 0 when the error occurs.

`ValueError` means that one of the GMMs had nan in `covars_` matrix after the M-step. This could've been caused by using an improper (non positive semi-definite) covariance matrix in the E-step.

**Comment by Randyjp** [comment]
I think fitting a GMMHMM is broken for the current version of hmmlearn. #78

**Comment by huiyuHarvey** [comment]
Hi, Randy. I saw that issue too. But right now all I need is the diagonal covariance matrix and that should be working fine.

**Comment by huiyuHarvey** [comment]
Both ways wouldn't work...

**Comment by superbobry** [comment]
Could you please add the new version of the code and elaborate a bit on what went wrong?

**Comment by huiyuHarvey** [comment]
@superbobry hi, the new code is as below dealing with multiple sequences.
```
def _init_hmm_para(self):
        
        self.pi_prior = np.array([2.0, 1.0, 1.0, 1.0, 1.0])
        self.A_prior = np.array([[1.6, 1.4, 1.0, 1.0, 1.0],
                                 [1.0, 1.5, 1.5, 1.0, 1.0],
                                 [1.0, 1.0, 1.4, 1.6, 1.0],
                                 [1.0, 1.0, 1.0, 1.5, 1.5],
                                 [1.0, 1.0, 1.0, 1.0, 2.0]])

        h = GMMHMM(self.n_node, self.n_gmm, covariance_type = 'diag',
            startprob_prior=self.pi_prior, transmat_prior=self.A_prior, 
            n_iter=self.n_iter, verbose=False, init_params='mcw',
            params='tmcw')
        h.startprob_ = self.pi_prior
        h.transmat_ = self.A_prior
        return h

    def train(self):
        h = self._init_hmm_para()
        flst = self._add_direct()
        len_cc = []
        X = [[0 for col in range(N_FEATURES)] for row in range(43)]
        for f in flst:
            print f
            audio = _input_audio(f)
            audio1 = _get_first_channel(audio)
            frames = _audio2frame(audio1)
            cc = _compute_cepstrum_coefficient(frames)
            len_cc.append(len(cc))
            X = np.concatenate([X, cc])
            
        h.fit(X, len_cc)
            
        pi = h.startprob_
        A = h.transmat_
        h = joblib.dump(h, "hmmgmm_model.pkl")
        
        return h, pi, A
```

At iteration 5 of first sequence, the `_accumulate_sufficient_statistics` of `hmm.py` raised a `ValueError`.

Could u plz explain how should I initialize a left-right GMMHMM model since each element of `transmat_` should be >= 1 to avoid negative value and any suggestions on rewriting the `normalize()` method of `utils.py` because each row of `transmat_` doesn't sum to 1 any more?

**Comment by huiyuHarvey** [comment]
what i do now is to initialize as below
```
self.A_prior = np.array([[1.8, 1.2, 0.0, 0.0, 0.0],
                         [0.0, 1.7, 1.3, 0.0, 0.0],
                         [0.0, 0.0, 1.6, 1.4, 0.0],
                         [0.0, 0.0, 0.0, 1.5, 1.5],
                         [0.0, 0.0, 0.0, 0.0, 2.0]])
```
and add 1 to the non-zero element in each iteration after normalization.
```
normalize(self.transmat_, axis=1)
self.transmat_ = np.where(self.transmat_ == 0.0,
                         self.transmat_, np.add(self.transmat_, 1))
```
but still got the error after a few iterations.

**Comment by huiyuHarvey** [comment]
I traced back the problem. @superbobry 
In the `_do_mstep()` of `hmm.py`, if the norm is 0, then will lead to `g.means_` to `Nan`
in line 635 `g.means_ = stats['means'][state] / norm[:, np.newaxis]`.

**Comment by superbobry** [comment]
Nice work @huiyuHarvey! Could you provide a complete example so we can test the `GMMHMM` from #107 against it?

**Comment by huiyuHarvey** [comment]
[X.zip](https://github.com/hmmlearn/hmmlearn/files/236563/X.zip)
This is the data for training which is numpy array containing 3 sequences in total.
The shape of X is `(n_samples=129, n_features=12)` 
`lengths` parameter is `[43, 43, 43]`
and initialization is as below:
```
pi_prior = np.array([2.0, 0.0, 0.0, 0.0, 0.0])
A_prior = np.array([[1.8, 1.2, 0.0, 0.0, 0.0],
                    [0.0, 1.7, 1.3, 0.0, 0.0],
                    [0.0, 0.0, 1.6, 1.4, 0.0],
                    [0.0, 0.0, 0.0, 1.5, 1.5],
                    [0.0, 0.0, 0.0, 0.0, 2.0]])
h = GMMHMM(n_components=5, n_mix=6, covariance_type='diag',
            startprob_prior=pi_prior, transmat_prior=A_prior, 
            n_iter=100, tol=0.1, verbose=False, init_params='mcw',
            params='tmcw')
```
It is left-right model. I followed your advice before to set the element greater than 1.
Let me know is anything else is needed. Hope it helps the rewrite of GMMHMM class.
